brave-intl/bat-go
# Fix chronic mapping issues in ratios service

### Summary

This PR fixes a bug in Ratios that has been plaguing us since at least the past 3 years. It brings about the following changes to Ratios service, regarding how we resolve a `coin` string into an appropriate CoinGecko ID.
1. Prioritise querying prices by CoinGecko ID vs symbol.
2. Fix the initialisation logic to correctly populate `idToSymbol` and `symbolToID` mappings.

#### What was the bug? 🐛 

It's possible to have multiple coins with the same symbol. This means when the `symbolToID` map is populated, the coin that's processed last will end up taking priority. CoinGecko keeps listing new coins with symbol collisions every now and then, so something that works today may not work tomorrow. Consider this example coin list:

```json
[
  {
    "id": "bitcoin",
    "symbol": "btc",
    "name": "Bitcoin",
  },
  {
    "id": "blackrocktradingcurrency",
    "symbol": "btc",
    "name": "BlackrockTradingCurrency"
  }
]
```

If you query prices by the string `btc`, it'll resolve to `blackrocktradingcurrency`, since it's going to supersede `bitcoin` in the `symbolToID` map.

Mappings like `bonk->bonk` are not required anymore, and therefore have been removed.

#### Note to clients

1. Clients should prioritise using CoinGecko IDs over symbols or contract addresses, since they are not guaranteed to be unique. Queries that rely on the symbol are not guaranteed to report accurate prices. This is already the case with Brave Desktop, iOS, and Android (WebUI).

2. Clients currently query native asset prices by symbol. This should be changed to CoinGecko ID to reduce our dependency on mappings. Querying by symbol will still remain an option for unknown network assets.

#### Verification

| Query string |Production| This PR | Reference | Remarks |
|-|:-:|:-:|-|-|
|**`btc`**| ❌ | ✅ | https://www.coingecko.com/en/coins/bitcoin | Added mapping |
|**`bitcoin`**| ❌ | ✅ | https://www.coingecko.com/en/coins/bitcoin | Query by ID |
|**`eth`**| ✅ | ✅ | https://www.coingecko.com/en/coins/ethereum | Uses existing mapping |
|**`ethereum`**| ❌ | ✅ | https://www.coingecko.com/en/coins/ethereum | Query by ID |
|**`tether`**| ✅ | ✅ | https://www.coingecko.com/en/coins/tether | Query by ID |
|**`usdt`**| ❌ | ✅ | https://www.coingecko.com/en/coins/tether | Added mapping |
|**`bat`**| ✅ | ✅ | https://www.coingecko.com/en/coins/basic-attention-token | Query by symbol |
|**`basic-attention-token`**| ✅ | ✅ | https://www.coingecko.com/en/coins/basic-attention-token | Query by ID |
|**`dai`**| ✅ | ✅ | https://www.coingecko.com/en/coins/dai| Works without a mapping |
|**`matic`**| ✅ | ✅ | https://www.coingecko.com/en/coins/matic-network| Query by symbol |
|**`matic-network`**| ✅ | ✅ | https://www.coingecko.com/en/coins/matic-network| Query by ID |
|**`thor`**| ✅ | ✅ | https://www.coingecko.com/en/coins/thor| Works without a mapping |
|**`bonk`**| ✅ | ✅ | https://www.coingecko.com/en/coins/bonk| Works without a mapping |
|**`ethereum-wormhole`**| ❌ | ✅ | https://www.coingecko.com/en/coins/ethereum-wormhole | Query by ID |
|**`staked-ether`**| ✅ | ✅ | https://www.coingecko.com/en/coins/lido-staked-ether | Query by ID |
|**`steth`**| ✅ | ✅ | https://www.coingecko.com/en/coins/lido-staked-ether | Query by symbol |

### Type of Change

- [ ] Product feature




### Tested Environments



### Before Requesting Review

- [x] Does your code build cleanly without any errors or warnings?
- [ ] Have you used auto closing keywords?
- [ ] Have you added tests for new functionality?
- [ ] Have validated query efficiency for new database queries?
- [x] Have documented new functionality in README or in comments?
- [ ] Have you squashed all intermediate commits?
- [x] Is there a clear title that explains what the PR does?
- [ ] Have you used intuitive function, variable and other naming?
- [ ] Have you requested security and/or privacy review if needed
- [x] Have you performed a self review of this PR?


### Manual Test Plan

Use the table under **Verification** head to verify if the prices are resolved accurately to the reference CoinGecko asset.
